packages/evm/src/opcodes/functions.ts

```diff
@@ -760,43 +761,91 @@ export const handlers: Map<number, OpHandler> = new Map([
   ],
   // 0x5b: JUMPDEST
   [0x5b, function () {}],
-  // 0x5c: BEGINSUB
+  // 0x5c: BEGINSUB (2315) or RJUMP (4200)
   [
     0x5c,
-    function (runState) {
-      trap(ERROR.INVALID_BEGINSUB + ' at ' + describeLocation(runState))
+    function (runState, common) {
+      if (common.isActivatedEIP(2315)) {
+        trap(ERROR.INVALID_BEGINSUB + ' at ' + describeLocation(runState))
+      } else if (common.isActivatedEIP(4200)) {
+        if (runState.env.containerCode![0] === EOF_FORMAT) {
+          const code = runState.env.code
+          const rjumpDest = code.readInt16BE(runState.programCounter)
+          runState.programCounter += 2 + rjumpDest
+        } else {
+          // Legacy contracts do not support RJUMP
+          trap(ERROR.INVALID_OPCODE)
+        }
+      }
     },
   ],
-  // 0x5d: RETURNSUB
+  // 0x5d: RETURNSUB (2315) or RJUMPI (4200)
   [
     0x5d,
-    function (runState) {
-      if (runState.returnStack.length < 1) {
-        trap(ERROR.INVALID_RETURNSUB)
-      }
+    function (runState, common) {
+      if (common.isActivatedEIP(2315)) {
+        if (runState.returnStack.length < 1) {
+          trap(ERROR.INVALID_RETURNSUB)
+        }
 
-      const dest = runState.returnStack.pop()
-      runState.programCounter = Number(dest)
+        const dest = runState.returnStack.pop()
+        runState.programCounter = Number(dest)
+      } else if (common.isActivatedEIP(4200)) {
```

Comment:
Oh, we've got a double opcode occupancy here? 🤔

Reply:
Yup. But the other EIP (Subroutines) seems to me a kind of predecessor to this EIP, which is also why most likely the same opcodes are choosen.